### PR TITLE
Fixes couple of native memory leaks

### PR DESCRIPTION
**Pull request by tonyredondo**
This PR is a continuation of https://github.com/DataDog/dd-trace-dotnet/pull/1596 handling a couple of possible memory leaks in the native library.

@DataDog/apm-dotnet

### CODE REVIEW

File: src/Datadog.Trace.ClrProfiler.Native/rejit_handler.cpp [hunk 1]
```diff
@@ -135,6 +134,18 @@ bool RejitHandler::TryGetModule(ModuleID moduleId, RejitHandlerModule** moduleHa
...
 }
 
+void RejitHandler::ReleaseModule(ModuleID moduleId)
+{
+    std::lock_guard<std::mutex> guard(modules_lock);
+    auto find_res = modules.find(moduleId);
+    if (find_res != modules.end())
+    {
+        auto moduleHandler = find_res->second;
+        delete[] moduleHandler;
```
Comment (gleocadie) [inline src/Datadog.Trace.ClrProfiler.Native/rejit_handler.cpp:144]: shoudn't it be `delete moduleHandler` instead ?

File: src/Datadog.Trace.ClrProfiler.Native/rejit_handler.h [hunk 2]
```diff
@@ -37,46 +40,58 @@ class RejitHandlerModuleMethod
...
         this->methodReplacement = nullptr;
     }
-    inline mdMethodDef GetMethodDef()
+    ~RejitHandlerModuleMethod()
+    {
+        this->pFunctionControl = nullptr;
+        this->module = nullptr;
+        if (this->functionInfo != nullptr)
+        {
+            delete[] this->functionInfo;
```
Comment (gleocadie) [inline src/Datadog.Trace.ClrProfiler.Native/rejit_handler.h:60]: shouldn't it be `delete this->functionInfo` instead ?

File: src/Datadog.Trace.ClrProfiler.Native/rejit_handler.h [hunk 3]
```diff
@@ -94,28 +109,38 @@ class RejitHandlerModule
...
         this->handler = handler;
     }
-    inline ModuleID GetModuleId()
+    ~RejitHandlerModule()
+    {
+        this->metadata = nullptr;
+        this->handler = nullptr;
+        for (auto moduleMethod : methods)
+        {
+            delete[] moduleMethod.second;
```
Comment (gleocadie) [inline src/Datadog.Trace.ClrProfiler.Native/rejit_handler.h:127]: should'nt it be `delete moduleMethod.second` instead?

File: src/Datadog.Trace.ClrProfiler.Native/rejit_handler.h [hunk 4]
```diff
@@ -151,9 +176,23 @@ class RejitHandler
         this->rejit_queue_ = new BlockingQueue<RejitItem>();
         this->rejit_queue_thread_ = new std::thread(enqueue_thread, this);
     }
+    ~RejitHandler()
+    {
+        if (this->rejit_queue_ != nullptr)
+        {
+            delete[] this->rejit_queue_;
```
Comment (gleocadie) [inline src/Datadog.Trace.ClrProfiler.Native/rejit_handler.h:183]: shoudn't it be `delete this->rejit_queue_;` instead?

File: src/Datadog.Trace.ClrProfiler.Native/rejit_handler.h [hunk 5]
```diff
@@ -151,9 +176,23 @@ class RejitHandler
...
+            this->rejit_queue_ = nullptr;
+        }
+        if (this->rejit_queue_thread_ != nullptr)
+        {
+            delete[] this->rejit_queue_thread_;
```
Comment (gleocadie) [inline src/Datadog.Trace.ClrProfiler.Native/rejit_handler.h:188]: shouldn't it be`delete this->rejit_queue_thread_;` instead ?

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp [hunk 6]
```diff
@@ -651,6 +655,8 @@ HRESULT STDMETHODCALLTYPE CorProfiler::Shutdown()
     if (rejit_handler != nullptr)
     {
         rejit_handler->Shutdown();
+        delete[] rejit_handler;
```
Comment (gleocadie) [inline src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp:658]: should'nt it be `delete rejit_handler;` instead ?

File: src/Datadog.Trace.ClrProfiler.Native/module_metadata.h [hunk 7]
```diff
@@ -48,6 +48,14 @@ class ModuleMetadata
         corAssemblyProperty(corAssemblyProperty)
     {
     }
+    ~ModuleMetadata()
+    {
+        if (calltargetTokens != nullptr)
+        {
+            delete[] calltargetTokens;
```
Comment (gleocadie) [inline src/Datadog.Trace.ClrProfiler.Native/module_metadata.h:55]: shouldn't it be `delete calltargetTokens` instead ?

**Review by gleocadie — CHANGES_REQUESTED** [review]
most of the deletion should be `delete` and not `delete[]`.
Rule being:
- Allocate using `new`, delete with `delete`
- Allocate using `new[]` delete with `delete[]`

The mixup is considered as undefined behavior

Maybe someday, as a big refactoring moving to `std::unique_ptr`/`std::shared_ptr` (depending on the semantic) could be a better  option.

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp [hunk 8]
```diff
@@ -3479,6 +3485,8 @@ HRESULT CorProfiler::CallTarget_RewriterCallback(RejitHandlerModule* moduleHandl
 
     hr = rewriter.Export();
 
+    delete[] newEHClauses;
```
Comment (gleocadie) [inline src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp:3488]: That could be risky no? you pass its pointer to the rewriter and ends it lifecyle before the rewriter lifecyle.
Couldn't you use `std::unique_ptr` instead ? (and passing the pointer to the rewriter using the `.get` method )?

File: src/Datadog.Trace.ClrProfiler.Native/module_metadata.h (hunk 7 above)
Comment (gleocadie) [reply src/Datadog.Trace.ClrProfiler.Native/module_metadata.h:55]: I think in this case you can go for a `std::unique_ptr`.

**Comment by tonyredondo** [comment]
> most of the deletion should be `delete` and not `delete[]`.
> Rule being:
> 
> * Allocate using `new`, delete with `delete`
> * Allocate using `new[]` delete with `delete[]`
> 
> The mixup is considered as undefined behavior
> 
> Maybe someday, as a big refactoring moving to `std::unique_ptr`/`std::shared_ptr` (depending on the semantic) could be a better option.

Yes is a WIP, I had to go out before completing this. That’s why the WIP label and the Draft 😊

File: src/Datadog.Trace.ClrProfiler.Native/rejit_handler.h (hunk 2 above)
Comment (gleocadie) [reply src/Datadog.Trace.ClrProfiler.Native/rejit_handler.h:60]: It's possible to use `std::unique_ptr` here instead.

File: src/Datadog.Trace.ClrProfiler.Native/rejit_handler.h (hunk 3 above)
Comment (gleocadie) [reply src/Datadog.Trace.ClrProfiler.Native/rejit_handler.h:127]: for `methods` it's also possible to use `std::unique_ptr`

File: src/Datadog.Trace.ClrProfiler.Native/rejit_handler.h [hunk 9]
```diff
@@ -151,9 +176,23 @@ class RejitHandler
         this->rejit_queue_ = new BlockingQueue<RejitItem>();
         this->rejit_queue_thread_ = new std::thread(enqueue_thread, this);
     }
+    ~RejitHandler()
```
Comment (gleocadie) [inline src/Datadog.Trace.ClrProfiler.Native/rejit_handler.h:179]: for both, it's possible to use std::unique_ptr instead

**Comment by gleocadie** [comment]
I tested a bit and overall, you can replace every raw pointers by a std::unique_ptr ;)

**Comment by tonyredondo** [comment]
Superseeded by #1600

**Pull request closed without merging**